prow/tide/tide.go

```diff
@@ -835,15 +836,38 @@ func (c *Controller) mergePRs(sp subpool, prs []PullRequest) error {
 	return nil
 }
 
+func aggregateChangeProvider(providers []config.ChangedFilesProvider) config.ChangedFilesProvider {
+	return func() ([]string, error) {
+		changes := sets.NewString()
+		var errs []error
+		for _, provider := range providers {
+			specificChanges, specificErr := provider()
+			changes.Insert(specificChanges...)
+			errs = append(errs, specificErr)
+		}
+		return changes.UnsortedList(), errorutil.NewAggregate(errs...)
+	}
+}
+
 func (c *Controller) trigger(sp subpool, presubmitContexts map[int]sets.String, prs []PullRequest) error {
 	requiredContexts := sets.NewString()
 	for _, pr := range prs {
 		requiredContexts = requiredContexts.Union(presubmitContexts[int(pr.Number)])
 	}
 
-	// TODO(cjwagner): DRY this out when generalizing triggering code (and code to determine required and to-run jobs).
 	for _, ps := range c.ca.Config().Presubmits[sp.org+"/"+sp.repo] {
-		if ps.SkipReport || !ps.RunsAgainstBranch(sp.branch) || !requiredContexts.Has(ps.Context) {
+		if !requiredContexts.Has(ps.Context) || !ps.ContextRequired() {
+			continue
+		}
+		var providers []config.ChangedFilesProvider
+		for _, pr := range prs {
+			providers = append(providers, config.NewGitHubDeferredChangedFilesProvider(c.ghc, sp.org, sp.repo, int(pr.Number)))
```

Comment:
If we keep this it needs to be updated to use the file change cache like `presubmitsByPull` does.